src/serviceContext.js

```
@@ -0,0 +1,202 @@
+      this.redirectUris = [null];
+    }
+
+    this.callback = this.config['callback'] || {};
+
+    if (config && Object.keys(config).indexOf('keydefs') !== -1) {
+      this.keyjar = this.keyjar.buildKeyJar(config['keydefs'], this.keyjar)[1];
+    }
+
+    return this;
+  }
+
+  getClientSecret() {
+    return this.client_secret;
+  }
+
+  setClientSecret(val) {
+    if (!val) {
+      this.client_secret = '';
+    } else {
+      this.client_secret = val;
+      // client uses it for signing
+      // Server might also use it for signing which means the
+      // client uses it for verifying server signatures
+      if (this.keyjar == null) {
+        this.keyjar = new KeyJar();
+      }
+      this.keyjar.addSymmetric('', val.toString());
+    }
+  }
+
+  /**
+   *  Need to generate a redirect_uri path that is unique for a OP/RP combo
+   *  This is to counter the mix-up attack.
+   * @param {string} path Leading path
+   * @return A list of one unique URL
+   */
+  generateRequestUris(path) {
+    let m = crypto.createHmac('sha256', '');
+    try {
+      m.update(this.providerInfo['issuer']);
+    } catch (error) {
+      m.update(this.issuer);
+    }
+    m.update(this.base_url);
+    if (!path.startsWith('/')){
+      return [this.base_url + '/' + path+ '/' +  m.digest('hex')];
+    }else{
+      return [this.base_url + path + '/' + m.digest('hex')];
+    }
+  }
+
+  /**
+   *  A 1<->1 map is maintained between a URL pointing to a file and
+   * the name of the file in the file system.
+   * 
+   * As an example if the base_url is 'https://example.com' and a jwks_uri
+   * is 'https://example.com/jwks_uri.json' then the filename of the
+   * corresponding file on the local filesystem would be 'jwks_uri.json.
+   * Relative to the directory from which the RP instance is run.
+   * 
+   * @param {string} webName 
+   */
+  filenameFromWebName(webName) {
+    if (webName.startsWith(this.base_url) == false){
+      console.log('ValueError');
```

I don't think that just logging and still proceeding afterwards is the right thing to do here.
Maybe raise an error here or at least do not proceed further, as it will then lead to unexpected behavior.